### PR TITLE
Implement pkg.uptodate for pkgin

### PR DESCRIPTION
**Pull request by sjorge**
### What does this PR do?
This PR implements ```list_upgrades``` for the pkgin provider for pkg.

This is needed in order to make the pkg.uptodate state work on SmartOS and NetBSD,
during the implementation I noticed pkgin.latest_version was broken in newer pkgin
in version 0.7.0 a new ```-p``` flag was introduced and that broke the parsing code.

```latest_version``` was fixed because ```list_upgrades``` uses the function internally.

### What issues does this PR fix or reference?
None, I just noticed this was broken in my setup.

### Previous Behavior
The pkg.uptodate state did not function with a newish pkgin

### New Behavior
The pkg.uptodate now works as expected

### Tests written?
No

### Commits signed with GPG?
No

@cachedout is GPG signing required? I played with this ages ago but could never get it safely working over ssh as my development machines are not near where I physically am and storing the key on the server does not seem safe.

### CODE REVIEW

**Comment by sjorge** [comment]
Oops found one more error:

```     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/opt/local/lib/python2.7/site-packages/salt/state.py", line 1843, in call
                  **cdata['kwargs'])
                File "/opt/local/lib/python2.7/site-packages/salt/loader.py", line 1795, in wrapper
                  return f(*args, **kwargs)
                File "/opt/local/lib/python2.7/site-packages/salt/states/pkg.py", line 2867, in uptodate
                  ret['changes'] = __salt__['pkg.upgrade'](refresh=refresh, pkgs=pkgs, **kwargs)
              TypeError: upgrade() takes no arguments (2 given)```

Changed title to [WIP] will fix this ASAP

**Comment by sjorge** [comment]
Fixed, first the easy way, then after chatting with one of the pkgin developers. The proper way :)

Calling install for an already install is actually and upgrade of a single package, this is also mentioned http://pkgin.net/.

This should now be ready, barring any issues found by jenkins or code review issues brought up.

**Comment by rallytime** [comment]
re-run ubuntu

**Comment by cachedout** [comment]
@sjorge GPG signing is not required. Just encouraged. Thanks.

File: salt/modules/pkgin.py [hunk 1]
```diff
@@ -81,10 +81,17 @@ def _supports_regex():
     '''
     Check support of regexp
     '''
-
     return tuple([int(i) for i in _get_version()]) > (0, 5)
 
 
+@decorators.memoize
+def _supports_parsing():
```
Comment (sjorge) [inline salt/modules/pkgin.py]: As per the change log in 0.7.0 the ```-p``` flag was introduced to
generate parse-able output.

We assume we have this output if we have the regex support but those are
separate features. This commit adds a new internal function to detect the
support. This also fixes latest_version to now work properly on systems
with this newer version of pkgin.

File: salt/modules/pkgin.py [hunk 2]
```diff
@@ -410,10 +450,25 @@ def install(name=None, refresh=False, fromrepo=None,
     return ret
 
 
-def upgrade():
+def upgrade(refresh=True, pkgs=None, **kwargs):
```
Comment (sjorge) [inline salt/modules/pkgin.py]: these extra flags are needed for the pkg.uptodate state.

**Comment by sjorge** [comment]
@rallytime can you take another look at this? Looks to still be the same IOLoop error in jenkins for ubuntu but I get that on all PR's

**Comment by rallytime** [comment]
Yeah, we had an upstream error that has been fixed. I'll update the branch and kick off a new run, just in case.

**Comment by sjorge** [comment]
Looks like jenkins is happy, aside from GPG!

File: salt/modules/pkgin.py [hunk 3]
```diff
@@ -98,6 +105,7 @@ def __virtual__():
 
 
 def _splitpkg(name):
+    '''Split package name from versioned string'''
```
Comment (rallytime) [inline salt/modules/pkgin.py]: Can you split this docstring into 3 separate lines?

File: salt/modules/pkgin.py [hunk 4]
```diff
@@ -214,7 +227,7 @@ def version(*names, **kwargs):
     return __salt__['pkg_resource.version'](*names, **kwargs)
 
 
-def refresh_db():
+def refresh_db(force=False):
     '''
     Use pkg update to get latest pkg_summary
```
Comment (rallytime) [inline salt/modules/pkgin.py]: Can you add documentation for the new `force` option here? (Don't forget to add a `.. versionadded:: Oxygen` tag, too. :)

File: salt/modules/pkgin.py [hunk 5]
```diff
@@ -295,6 +313,28 @@ def list_pkgs(versions_as_list=False, **kwargs):
     return ret
 
 
+def list_upgrades(refresh=True, **kwargs):
+    '''
+    List all available package upgrades.
+
```
Comment (rallytime) [inline salt/modules/pkgin.py]: This could use a `.. versionadded:: Oxygen` tag here for this new function.

**Review by rallytime — CHANGES_REQUESTED** [review]
@sjorge I have a couple of small comments, but otherwise this looks good to me.

File: salt/modules/pkgin.py (hunk 3 above)
Comment (sjorge) [reply salt/modules/pkgin.py]: Done, I would have expected lint to catch these as wel though?

File: salt/modules/pkgin.py (hunk 4 above)
Comment (sjorge) [reply salt/modules/pkgin.py]: Done

File: salt/modules/pkgin.py (hunk 5 above)
Comment (sjorge) [reply salt/modules/pkgin.py]: And done

**Comment by cachedout** [comment]
@rallytime Re-review please.

**Review by rallytime — APPROVED** [review]

**Pull request merged**